### PR TITLE
revert: "revert: "feat: support stopping training in trial code [DET-3238] (#648)" (#654)"

### PR DESCRIPTION
**Pull request by brain-good**
## Test Plan
- [x] gpu tests: https://app.circleci.com/pipelines/github/determined-ai/determined/2179/workflows/97b047ac-5d47-4c57-96ed-b587a3ef071c/jobs/43388/steps

## Commentary (optional)
This is the same PR as #648 but with fixes for dtrain.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/keras/_tf_keras_trial.py [hunk 1]
```diff
@@ -485,11 +485,11 @@ def run(self) -> None:
                 break
 
             elif wkld.kind == workload.Workload.Kind.COMPUTE_VALIDATION_METRICS:
-                response = {
-                    "metrics": self.compute_validation_metrics(),
-                    "stop_requested": self.context.get_stop_requested(),
-                }
-                response_func(response)
+                response_func(
+                    det.util.wrap_metrics(
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_trial.py:489]: nit: `util.wrap_metrics(`

File: harness/determined/tensorpack/_tensorpack_trial.py [hunk 2]
```diff
@@ -249,11 +249,11 @@ def _control_loop(self) -> None:
                 self.train_response_func = response_func
                 break
             elif wkld.kind == workload.Workload.Kind.COMPUTE_VALIDATION_METRICS:
-                response = {
-                    "metrics": self._compute_validation_metrics(),
-                    "stop_requested": self.context.get_stop_requested(),
-                }
-                response_func(response)
+                response_func(
+                    det.util.wrap_metrics(
```
Comment (aaron276h) [inline harness/determined/tensorpack/_tensorpack_trial.py:253]: ditto

File: harness/determined/_train_context.py [hunk 3]
```diff
@@ -82,6 +83,22 @@ def get_hparam(self, name: str) -> Any:
...
+        return self._stop_requested
+
+    def set_stop_requested(self, stop_requested: bool) -> None:
+        """
+        Set a flag to request a trial stoppage. When this flag is set to True,
+        we finish the step, checkpoint, then exit.
+        """
+        if not isinstance(stop_requested, bool):
+            raise AssertionError("stop_requested must be a boolean")
+
```
Comment (aaron276h) [inline harness/determined/_train_context.py:99]: can you add a log message here

File: harness/determined/keras/_tf_keras_trial.py (hunk 1 above)
Comment (brain-good) [reply harness/determined/keras/_tf_keras_trial.py:489]: gonna hold off --  we already use `det.util` in other places 👀

**Pull request merged**